Apps/System/music.c

```diff
@@ -0,0 +1,461 @@
+};
+
+static Animation *s_animation_record_ptr;
+static Animation *s_animation_arm_ptr;
+static bool s_animating_disk_change;
+static bool s_animating_arm_change;
+static bool s_is_paused;
+static bool s_was_paused;
+static char *s_artist;
+static char *s_track;
+static GBitmap *s_up_bitmap;
+static GBitmap *s_down_bitmap;
+static GBitmap *s_select_bitmap;
+static GColor s_curr_disk_color;
+static GColor s_next_disk_color;
+static GPoint s_curr_disk_pos;
+static GPoint s_next_disk_pos;
+static int32_t s_progress;
+static int32_t s_length;
+static int32_t s_progress_pixels;
+static int32_t s_skip_value;
+static int32_t s_arm_angle;
+static int32_t s_old_arm_angle;
+static int32_t s_animation_start_arm_angle;
+static int32_t s_animation_end_arm_angle;
+static struct tm s_last_time;
+// TODO remove these variables, they are for demoing only
+static int32_t s_trackcounter;
+static char *artists[] = { "The Beatles", "Deadmau5", "Daniel Ingram" };
+static char *tracks[] = { "Maxwell's Silver Hammer", "Strobe", "Smile" };
+static int32_t length[] = { 207, 637, 203 }; // { 46, 5, 10 };
+
+static void get_new_track(){
+    s_trackcounter++;
+    if(s_trackcounter >= 3) {
+        s_trackcounter = 0;
+    }
+    s_artist = artists[s_trackcounter];
+    s_track = tracks[s_trackcounter];
+    s_length = length[s_trackcounter];
+    s_progress = 0;
+}
+static void get_last_track(){
+    s_trackcounter--;
+    if(s_trackcounter < 0) {
+        s_trackcounter = 2;
+    }
+    s_artist = artists[s_trackcounter];
+    s_track = tracks[s_trackcounter];
+    s_length = length[s_trackcounter];
+    s_progress = 0;
+}
+
+static void music_tick(struct tm *tick_time, TimeUnits tick_units) {
+    if((tick_units & SECOND_UNIT) != 0 && !s_is_paused) {
+        if(s_progress >= s_length) {
+            get_new_track();
+            skip_track(SKIP_DIRECTION_NEXT);
+            s_progress_pixels = 0;
+        } else {
+            s_progress++;
+            s_progress_pixels = s_progress * (PROGRESS_PIXELS_MAX - 1) / s_length;
+            int32_t new_arm_angle = ARM_START_ANGLE + s_progress * (ARM_END_ANGLE - ARM_START_ANGLE) / s_length;
+            if(new_arm_angle != s_arm_angle){
+                animation_arm(new_arm_angle, ARM_SKIP_SPEED);
+            } else {
+                layer_mark_dirty(s_main_layer);
+            }
+        }
+    }
+    if((tick_units & MINUTE_UNIT) != 0) {
+        // Store time for status bar
+        memcpy(&s_last_time, tick_time, sizeof(struct tm));
+        layer_mark_dirty(s_main_layer);
+    }
+}
+
+static void implementation_record_setup(Animation *animation) {
+    s_animating_disk_change = true;
+    // TODO set next_disk_color based on title + artist,
+    // to let each track have the same color between sessions
+    // Avoid color black and white (Changing the stroke color would be an alternative)
+    s_next_disk_color =  (GColor8) { .argb = ((rand() % 0b00111111) + 0b11000000)  };
```

Comment:
This is a super cool trick.

Reply:
Found it somewhere online. 😄

Reply:
Small caveat: if rand() returns `0b00111111` the formula would return `0b11000000` (black).
If you set your dividend to `0b001111110` and the addend to `0b11000001` you should get your desired range of `0b11000001` (not quite black) to `0b11111110` (not quite white).